build: patch qt to explicitly define previously implicit header include

kwvg opened the pull request:
### Description

macOS Monterey has refactored some includes such that implicitly defined headers were no longer exposed and that in turns breaks building Qt on macOS 12. First observed when building on an M1 Pro. See "Additional Reading" for more information.

### Additional Reading

* https://github.com/microsoft/vcpkg/issues/21055 (resolved by https://github.com/microsoft/vcpkg/commit/c32ccbb34caacea28ba2d2b3df232ee36359c9f6)
* ["Add missing macOS header file that was indirectly included before"](https://code.qt.io/cgit/qt/qtbase.git/commit/src/plugins/platforms/cocoa?id=dece6f5840463ae2ddf927d65eb1b3680e34a547) on https://code.qt.io/

katesalazar commented:
Concept ACK.

fanquake commented:
Note that this patch is already part of https://github.com/bitcoin/bitcoin/pull/23489.

fanquake commented on depends/patches/qt/fix_montery_include.patch:
```diff
@@ -0,0 +1,11 @@
+diff --git a/qtbase/src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h b/qtbase/src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h
```
Can you add a comment here, the same as our other qt patches. Linking to the upstream commit.

kittywhiskers replied:
Done!

fanquake reviewed (commented):
Even though this is included in #23489, it will need to be backported separately to the 22.x branch. Please use a more informative commit message. i.e `build: patch qt in depends to fix missing header include`, and your PR description sections as the commit description.

kwvg committed:
build: patch qt to explicitly define previously implicit header include

macOS Monterey has refactored some includes such that implicitly defined headers were no longer exposed and that in turns breaks building Qt on macOS 12.

Additional Resources:
 - https://bugreports.qt.io/browse/QTBUG-97855
 - https://codereview.qt-project.org/c/qt/qtbase/+/378706
 - https://code.qt.io/cgit/qt/qtbase.git/commit/src/plugins/platforms/cocoa?id=dece6f5840463ae2ddf927d65eb1b3680e34a547

hebasto reviewed (approved):
ACK 8196b0a2bc63c35769eca213cba4a5844a737b90, verified that the patch is applied cleanly:
```
$ make -C depends qt_preprocessed
...
Preprocessing qt...
patching file qtbase/configure
patching file qtbase/mkspecs/features/qt_module.prf
patching file qtbase/src/plugins/platforms/cocoa/qprintengine_mac_p.h
patching file qtbase/src/plugins/plugins.pro
patching file qtbase/mkspecs/android-clang/qmake.conf
patching file qtbase/mkspecs/common/android-base-head.conf
patching file qtbase/mkspecs/common/android-base-tail.conf
patching file qtbase/src/plugins/platforms/android/androidjnimain.cpp
patching file qtbase/mkspecs/common/android-base-head.conf
patching file qtbase/src/plugins/platforms/xcb/qxcbcursor.cpp
patching file qtbase/mkspecs/features/mac/default_post.prf
patching file qtbase/mkspecs/common/mac.conf
patching file qtbase/mkspecs/features/qmake_use.prf
patching file qtbase/mkspecs/features/qt_configure.prf
patching file qtbase/src/tools/moc/main.cpp
patching file qtbase/src/corelib/global/qendian.h
patching file qtbase/src/corelib/tools/qbytearraymatcher.h
patching file qtbase/src/tools/moc/generator.cpp
patching file qtbase/src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h
make: Leaving directory '/home/hebasto/GitHub/bitcoin/depend
```

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
0e05d664a4d31bfa06c435ebcd54dd7ad931f54bd4e1f0d7ee1e656cfd838e8d  guix-build-8196b0a2bc63/output/aarch64-linux-gnu/SHA256SUMS.part
a5b582ebd605afd26cfd3fdae7c3da80da397db8d499049e53bbb9c2e9e9bf2c  guix-build-8196b0a2bc63/output/aarch64-linux-gnu/bitcoin-8196b0a2bc63-aarch64-linux-gnu-debug.tar.gz
d3047e352ed4902ca3d6cb006094c3deeb6d4b24cc19716d4f525a123f093856  guix-build-8196b0a2bc63/output/aarch64-linux-gnu/bitcoin-8196b0a2bc63-aarch64-linux-gnu.tar.gz
e48ac638316c738bab3cbb3d58b22e286f3df0f4e1865352b8503ad37424e984  guix-build-8196b0a2bc63/output/arm-linux-gnueabihf/SHA256SUMS.part
be48a254ad97a09a0274d58c4bcb7db0ff5abc6ac9b9c8ee6ad44c15d611a581  guix-build-8196b0a2bc63/output/arm-linux-gnueabihf/bitcoin-8196b0a2bc63-arm-linux-gnueabihf-debug.tar.gz
5125f5d9ad51116f2d98691d82b73e8a40bc8fb6152e59bc944ea4cf0897d130  guix-build-8196b0a2bc63/output/arm-linux-gnueabihf/bitcoin-8196b0a2bc63-arm-linux-gnueabihf.tar.gz
6369fb5d9bf94df26a469c8ab9455cecb30b3a7c45ddb117be8e24d6189a37b4  guix-build-8196b0a2bc63/output/dist-archive/bitcoin-8196b0a2bc63.tar.gz
be66c7dd8a77b60795ddcd794eb408b727ecd2c07a5801506d533866ab636a68  guix-build-8196b0a2bc63/output/powerpc64-linux-gnu/SHA256SUMS.part
272441042f07f60eea66cd3329bcb4da6c22787dca480e9ccc36c21982813ca6  guix-build-8196b0a2bc63/output/powerpc64-linux-gnu/bitcoin-8196b0a2bc63-powerpc64-linux-gnu-debug.tar.gz
595f213ed5ca384df8a0902dcce70db7bbfdee3bb401b101ad480579b0bf8f34  guix-build-8196b0a2bc63/output/powerpc64-linux-gnu/bitcoin-8196b0a2bc63-powerpc64-linux-gnu.tar.gz
1abeb3645c4a4b764cad0c2cf9fa5098472b16238027ee35e523ced415495503  guix-build-8196b0a2bc63/output/powerpc64le-linux-gnu/SHA256SUMS.part
be301d6de1a44a6db17da0219f374ef6ea92095cf643ff2d4dfd1601ba44cacf  guix-build-8196b0a2bc63/output/powerpc64le-linux-gnu/bitcoin-8196b0a2bc63-powerpc64le-linux-gnu-debug.tar.gz
bcbded980aad641b010f16324b5dd394119dba223ee076df7efe0896f6b8d477  guix-build-8196b0a2bc63/output/powerpc64le-linux-gnu/bitcoin-8196b0a2bc63-powerpc64le-linux-gnu.tar.gz
51d71b9216b857dfff81a28e8cc2fc7a40569edeec0af54c268c640a5e1f62b0  guix-build-8196b0a2bc63/output/riscv64-linux-gnu/SHA256SUMS.part
199b9586ab2b924a9f9519a91399a133e6d479fcba5a70a82f20355a3220314e  guix-build-8196b0a2bc63/output/riscv64-linux-gnu/bitcoin-8196b0a2bc63-riscv64-linux-gnu-debug.tar.gz
13f8787c17815f6d1fefccf11ea0008bfbe9c20662663ab01ac47829813c7735  guix-build-8196b0a2bc63/output/riscv64-linux-gnu/bitcoin-8196b0a2bc63-riscv64-linux-gnu.tar.gz
1bc7434433cc7276c45b50c9b6235f15b1a174b43b164d921a91c9bfe72bdc62  guix-build-8196b0a2bc63/output/x86_64-apple-darwin19/SHA256SUMS.part
d06073a444a587615b6b908333ea017c354a0bad965f35c144661a1b03128613  guix-build-8196b0a2bc63/output/x86_64-apple-darwin19/bitcoin-8196b0a2bc63-osx-unsigned.dmg
ddcfa4de1773bf4ca48ebebb8cf0cece119f559ac6e1df52f96a27714cfc3023  guix-build-8196b0a2bc63/output/x86_64-apple-darwin19/bitcoin-8196b0a2bc63-osx-unsigned.tar.gz
a9c6005d7034cfd9a5db0a36a659a07a1dd41c067e9b568391aeb08c9dbeae0a  guix-build-8196b0a2bc63/output/x86_64-apple-darwin19/bitcoin-8196b0a2bc63-osx64.tar.gz
72394cd581cc24df7c86ecedc8d291bcc1f83f8fb65b441d086e4e4da4371cdb  guix-build-8196b0a2bc63/output/x86_64-linux-gnu/SHA256SUMS.part
3001a2e7f3d2a6d3c9f98623647ddb98455618c7cd4e85d153b053c53c950037  guix-build-8196b0a2bc63/output/x86_64-linux-gnu/bitcoin-8196b0a2bc63-x86_64-linux-gnu-debug.tar.gz
156cf1e8e3e64890e9204280234b6c5f68da3e27637a67e14caf88111cd0f3fc  guix-build-8196b0a2bc63/output/x86_64-linux-gnu/bitcoin-8196b0a2bc63-x86_64-linux-gnu.tar.gz
faa104865f50c754c289f66ec49e47a451b643bcff581a37fafc64e91689c018  guix-build-8196b0a2bc63/output/x86_64-w64-mingw32/SHA256SUMS.part
7973638f2730b8945d84ff2f1448f136b273cb68d04ed22e52d50900446070b5  guix-build-8196b0a2bc63/output/x86_64-w64-mingw32/bitcoin-8196b0a2bc63-win-unsigned.tar.gz
bcc4fa825285518208758280f6479e7af67ece45c9619ce10f65aa50078c7100  guix-build-8196b0a2bc63/output/x86_64-w64-mingw32/bitcoin-8196b0a2bc63-win64-debug.zip
2e73dc1a535d6ae622f13485c1f6d596485ffdfbdeb70ce158a5824ae3b3fede  guix-build-8196b0a2bc63/output/x86_64-w64-mingw32/bitcoin-8196b0a2bc63-win64-setup-unsigned.exe
9c4ee4c09c7c3703b18edc59d9939148341ba5c80ba9b987f34fd452d3be8952  guix-build-8196b0a2bc63/output/x86_64-w64-mingw32/bitcoin-8196b0a2bc63-win64.zip
```

fanquake reviewed (approved):
ACK 8196b0a2bc63c35769eca213cba4a5844a737b90

Guix Build:
```bash
bash-5.1# find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
0e05d664a4d31bfa06c435ebcd54dd7ad931f54bd4e1f0d7ee1e656cfd838e8d  guix-build-8196b0a2bc63/output/aarch64-linux-gnu/SHA256SUMS.part
a5b582ebd605afd26cfd3fdae7c3da80da397db8d499049e53bbb9c2e9e9bf2c  guix-build-8196b0a2bc63/output/aarch64-linux-gnu/bitcoin-8196b0a2bc63-aarch64-linux-gnu-debug.tar.gz
d3047e352ed4902ca3d6cb006094c3deeb6d4b24cc19716d4f525a123f093856  guix-build-8196b0a2bc63/output/aarch64-linux-gnu/bitcoin-8196b0a2bc63-aarch64-linux-gnu.tar.gz
e48ac638316c738bab3cbb3d58b22e286f3df0f4e1865352b8503ad37424e984  guix-build-8196b0a2bc63/output/arm-linux-gnueabihf/SHA256SUMS.part
be48a254ad97a09a0274d58c4bcb7db0ff5abc6ac9b9c8ee6ad44c15d611a581  guix-build-8196b0a2bc63/output/arm-linux-gnueabihf/bitcoin-8196b0a2bc63-arm-linux-gnueabihf-debug.tar.gz
5125f5d9ad51116f2d98691d82b73e8a40bc8fb6152e59bc944ea4cf0897d130  guix-build-8196b0a2bc63/output/arm-linux-gnueabihf/bitcoin-8196b0a2bc63-arm-linux-gnueabihf.tar.gz
6369fb5d9bf94df26a469c8ab9455cecb30b3a7c45ddb117be8e24d6189a37b4  guix-build-8196b0a2bc63/output/dist-archive/bitcoin-8196b0a2bc63.tar.gz
be66c7dd8a77b60795ddcd794eb408b727ecd2c07a5801506d533866ab636a68  guix-build-8196b0a2bc63/output/powerpc64-linux-gnu/SHA256SUMS.part
272441042f07f60eea66cd3329bcb4da6c22787dca480e9ccc36c21982813ca6  guix-build-8196b0a2bc63/output/powerpc64-linux-gnu/bitcoin-8196b0a2bc63-powerpc64-linux-gnu-debug.tar.gz
595f213ed5ca384df8a0902dcce70db7bbfdee3bb401b101ad480579b0bf8f34  guix-build-8196b0a2bc63/output/powerpc64-linux-gnu/bitcoin-8196b0a2bc63-powerpc64-linux-gnu.tar.gz
1abeb3645c4a4b764cad0c2cf9fa5098472b16238027ee35e523ced415495503  guix-build-8196b0a2bc63/output/powerpc64le-linux-gnu/SHA256SUMS.part
be301d6de1a44a6db17da0219f374ef6ea92095cf643ff2d4dfd1601ba44cacf  guix-build-8196b0a2bc63/output/powerpc64le-linux-gnu/bitcoin-8196b0a2bc63-powerpc64le-linux-gnu-debug.tar.gz
bcbded980aad641b010f16324b5dd394119dba223ee076df7efe0896f6b8d477  guix-build-8196b0a2bc63/output/powerpc64le-linux-gnu/bitcoin-8196b0a2bc63-powerpc64le-linux-gnu.tar.gz
51d71b9216b857dfff81a28e8cc2fc7a40569edeec0af54c268c640a5e1f62b0  guix-build-8196b0a2bc63/output/riscv64-linux-gnu/SHA256SUMS.part
199b9586ab2b924a9f9519a91399a133e6d479fcba5a70a82f20355a3220314e  guix-build-8196b0a2bc63/output/riscv64-linux-gnu/bitcoin-8196b0a2bc63-riscv64-linux-gnu-debug.tar.gz
13f8787c17815f6d1fefccf11ea0008bfbe9c20662663ab01ac47829813c7735  guix-build-8196b0a2bc63/output/riscv64-linux-gnu/bitcoin-8196b0a2bc63-riscv64-linux-gnu.tar.gz
1bc7434433cc7276c45b50c9b6235f15b1a174b43b164d921a91c9bfe72bdc62  guix-build-8196b0a2bc63/output/x86_64-apple-darwin19/SHA256SUMS.part
d06073a444a587615b6b908333ea017c354a0bad965f35c144661a1b03128613  guix-build-8196b0a2bc63/output/x86_64-apple-darwin19/bitcoin-8196b0a2bc63-osx-unsigned.dmg
ddcfa4de1773bf4ca48ebebb8cf0cece119f559ac6e1df52f96a27714cfc3023  guix-build-8196b0a2bc63/output/x86_64-apple-darwin19/bitcoin-8196b0a2bc63-osx-unsigned.tar.gz
a9c6005d7034cfd9a5db0a36a659a07a1dd41c067e9b568391aeb08c9dbeae0a  guix-build-8196b0a2bc63/output/x86_64-apple-darwin19/bitcoin-8196b0a2bc63-osx64.tar.gz
72394cd581cc24df7c86ecedc8d291bcc1f83f8fb65b441d086e4e4da4371cdb  guix-build-8196b0a2bc63/output/x86_64-linux-gnu/SHA256SUMS.part
3001a2e7f3d2a6d3c9f98623647ddb98455618c7cd4e85d153b053c53c950037  guix-build-8196b0a2bc63/output/x86_64-linux-gnu/bitcoin-8196b0a2bc63-x86_64-linux-gnu-debug.tar.gz
156cf1e8e3e64890e9204280234b6c5f68da3e27637a67e14caf88111cd0f3fc  guix-build-8196b0a2bc63/output/x86_64-linux-gnu/bitcoin-8196b0a2bc63-x86_64-linux-gnu.tar.gz
faa104865f50c754c289f66ec49e47a451b643bcff581a37fafc64e91689c018  guix-build-8196b0a2bc63/output/x86_64-w64-mingw32/SHA256SUMS.part
7973638f2730b8945d84ff2f1448f136b273cb68d04ed22e52d50900446070b5  guix-build-8196b0a2bc63/output/x86_64-w64-mingw32/bitcoin-8196b0a2bc63-win-unsigned.tar.gz
bcc4fa825285518208758280f6479e7af67ece45c9619ce10f65aa50078c7100  guix-build-8196b0a2bc63/output/x86_64-w64-mingw32/bitcoin-8196b0a2bc63-win64-debug.zip
2e73dc1a535d6ae622f13485c1f6d596485ffdfbdeb70ce158a5824ae3b3fede  guix-build-8196b0a2bc63/output/x86_64-w64-mingw32/bitcoin-8196b0a2bc63-win64-setup-unsigned.exe
9c4ee4c09c7c3703b18edc59d9939148341ba5c80ba9b987f34fd452d3be8952  guix-build-8196b0a2bc63/output/x86_64-w64-mingw32/bitcoin-8196b0a2bc63-win64.zip
```

fanquake commented:
Added to #23276 for backport to 22.x.
